src/v/raft/replicate_entries_stm.cc

```diff
@@ -153,10 +153,24 @@ replicate_entries_stm::append_to_self() {
           return result<storage::append_result>(errc::leader_append_failed);
       });
 }
-
-inline bool replicate_entries_stm::is_follower_recovering(vnode id) {
+/**
+ *  We skip sending follower requests it those two cases:
+ *   - follower is recovering - when follower is not fully caught up it will not
+ *     accept append entries request, missing data will be replicated to
+ *     follower during recovery process
+ *   - we haven't received any response from the follower for replicate append
+ *     timeout duration - follower is probably down, we will not be able to
+ *     send the request to the follower and it will require recovery. This we
+ *     prevent pending follower request queue build up and relieve memory
+ *     pressure.
+ */
+inline bool replicate_entries_stm::should_skip_follower_request(vnode id) {
     if (auto it = _ptr->_fstats.find(id); it != _ptr->_fstats.end()) {
-        return it->second.is_recovering;
+        const auto timeout = clock_type::now()
+                             - _ptr->_replicate_append_timeout;
+
+        return it->second.last_hbeat_timestamp < timeout
```

Comment:
is it guaranteed that when the timeout here causes further follower requests to be skipped that some other mechanism then becomes responsible for flipping the switch back? that would seem to be that we receive a heartbeat from the follower, but is receiving that heartbeat _dependent_ on the request we are now skipping? maybe clarifying that bit of protocol magic here would be useful.

Reply:
Heartbeats are send with separate code path, they will not be skipped. After the follower will receive heartbeat the switch will be flipped back. I've added comment describing this.